database/src/backend/heed/database.rs

```diff
@@ -49,6 +49,17 @@ pub(super) struct HeedTableRw<'env, 'tx, T: Table> {
     pub(super) tx_rw: &'tx RefCell<heed::RwTxn<'env>>,
 }
 
+/// SAFETY: `cuprate_database`'s Cargo.toml enables a feature
+/// for `heed` that turns on the `MDB_NOTLS` flag for LMDB.
+///
+/// This makes read transactions `Send`, but only if that flag is enabled.
+///
+/// This is required as in `crate::service` we must put our transactions and
+/// tables inside `ThreadLocal`'s to use across multiple threads.
+///
+/// `ThreadLocal<T>` requires that `T: Send`.
+unsafe impl<T: Table> Send for HeedTableRo<'_, T> {}
```

Comment:
This is safe?

The feature makes `heed::RoTxn: Send`, and I believe `heed::Database` is `Send` for all of our cases.

Reply:
This is making `heed::RoTxn` `Sync`, which I don't think is safe.

For our current use cases this is ok as we aren't using them between threads however this would be a soundness hole. 

I only really see 2 ways of making this sound: add a mutex on the inner tx or mark every method on `HeedTableRo` unsafe. Both are not great.

Reply:
We're using `read-txn-no-tls` which only allows `Send`. `sync-read-txn` is removed in the version of `heed` we're using. 
 
![image](https://github.com/Cuprate/cuprate/assets/101352116/f741a78b-36aa-4079-bd36-b1d320b4e3df)

![image](https://github.com/Cuprate/cuprate/assets/101352116/12963568-b2c3-4e21-b04c-57693cf6889c)

Sidenote but this would be a good excuse to add `static_assertions` so we assert traits (among other things).

Reply:
I should have been more clear, but this is what `HeedTableRo` looks like:

```rust
pub(super) struct HeedTableRo<'tx, T: Table> {
    /// An already opened database table.
    pub(super) db: HeedDb<T::Key, T::Value>,
    /// The associated read-only transaction that opened this table.
    pub(super) tx_ro: &'tx heed::RoTxn<'tx>,
}
```

So impl'ing `Send` for `HeedTableRo` means you are making `heed::RoTxn` `Sync` (in this struct). You could create 2 `HeedTableRo` from the same tx and send one to another thread, then you would have the same tx shared between 2 threads.

Reply:
Ahhh I see, `&T: Send` only if `T: Sync`. I think Mutex is how `redb` does it?

I know this is rough but in our case (non-contended mutex) it's 3-5% slower. I think it's okay unless you preferred something else, although this would allow us to `impl Sync` for our tables too? No more `ThreadLocal`?

```rust
use heed::{types::*, *};
use std::{hint::black_box, sync::Mutex, time::Instant};

fn main() {
    let dir = tempfile::tempdir().unwrap();
    let env = unsafe { EnvOpenOptions::new().open(dir.path()).unwrap() };

    let mut tx_rw = env.write_txn().unwrap();
    let db: Database<Str, Str> = env.create_database(&mut tx_rw, None).unwrap();

    db.put(&mut tx_rw, "a", "").unwrap();
    tx_rw.commit().unwrap();

    const RANGE: std::ops::Range<usize> = 0..10_000_000;

    loop {
        let now = Instant::now();
        for _ in RANGE {
            let tx_ro = black_box(Mutex::new(env.read_txn().unwrap()));
            let _get = black_box(db.get(&tx_ro.lock().unwrap(), "zero")).unwrap();
        }
        println!("mutex: {}", now.elapsed().as_secs_f32());

        let now = Instant::now();
        for _ in RANGE {
            let tx_ro = black_box(env.read_txn()).unwrap();
            let _get = black_box(db.get(&tx_ro, "zero")).unwrap();
        }
        println!("tx_ro: {}", now.elapsed().as_secs_f32());
    }
}

```
```
mutex: 0.8341315
tx_ro: 0.76794577
mutex: 0.7981665
tx_ro: 0.7648049
mutex: 0.7917462
tx_ro: 0.76621574
```

Reply:
> although this would allow us to impl Sync for our tables too? No more ThreadLocal?

We would then have contention between threads retrieving data, only one will be able to at a time.

>  I think Mutex is how redb does it

Does this mean only one thread can get data an once? if so I would recommend putting that in a thread local as well 

>  3-5% slower

Not awful, still annoying though, probably our best option